rust/src/dataset/fragment.rs

```diff
@@ -92,20 +137,138 @@ impl FileFragment {
         Scanner::from_fragment(self.dataset.clone(), self.metadata.clone())
     }
+
+    /// Create an [`Updater`] to append new columns.
+    pub async fn updater<T: AsRef<str>>(&self, columns: Option<&[T]>) -> Result<Updater> {
+        let mut schema = self.dataset.schema().clone();
+        if let Some(columns) = columns {
+            schema = schema.project(&columns)?;
+        }
+        let reader = self.open(&schema).await?;
+
+        Ok(Updater::new(self.clone(), reader))
+    }
 }
 
-#[cfg(test)]
-mod tests {
+/// [`FragmentReader`] is an abstract reader for a [`FileFragment`].
+///
+/// It opens the data files that contains the columns of the projection schema, and
+/// reconstruct the RecordBatch from columns read from each data file.
+pub struct FragmentReader {
+    /// Readers and schema of each opened data file.
+    readers: Vec<(FileReader, Schema)>,
 
-    use crate::{arrow::RecordBatchBuffer, dataset::WriteParams};
+    /// ID of the fragment
+    fragment_id: usize,
+}
 
-    use super::*;
+impl std::fmt::Display for FragmentReader {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "FragmentReader(id={})", self.fragment_id)
+    }
+}
+
+fn merge_batches(batches: &[RecordBatch]) -> Result<RecordBatch> {
+    if batches.is_empty() {
+        return Err(Error::IO("Cannot merge empty batches".to_string()));
+    }
+
+    let mut merged = batches[0].clone();
+    for i in 1..batches.len() {
+        merged = merged.merge(&batches[i])?;
+    }
+    Ok(merged)
+}
+
+impl FragmentReader {
+    fn try_new(fragment_id: usize, readers: Vec<(FileReader, Schema)>) -> Result<Self> {
+        if readers.is_empty() {
+            return Err(Error::IO(
+                "Cannot create FragmentReader with zero readers".to_string(),
+            ));
+        }
+
+        let num_batches = readers[0].0.num_batches();
+        if !readers.iter().all(|r| r.0.num_batches() == num_batches) {
+            return Err(Error::IO(
+                "Cannot create FragmentReader from data files with different number of batches"
+                    .to_string(),
+            ));
+        }
+        Ok(Self {
+            readers,
+            fragment_id,
+        })
+    }
+
+    pub(crate) fn with_row_id(&mut self) -> &mut Self {
+        self.readers[0].0.with_row_id(true);
+        self
+    }
+
+    pub(crate) fn num_batches(&self) -> usize {
+        self.readers[0].0.num_batches()
+    }
+
+    pub(crate) fn num_rows_in_batch(&self, batch_id: usize) -> usize {
+        self.readers[0].0.num_rows_in_batch(batch_id as i32)
+    }
+
+    pub(crate) async fn read_batch(
+        &self,
+        batch_id: usize,
+        params: impl Into<ReadBatchParams> + Clone,
+    ) -> Result<RecordBatch> {
+        // TODO: use tokio::async buffer to make parallel reads.
+        let mut batches = vec![];
+        for (reader, schema) in self.readers.iter() {
+            let batch = reader
```

Comment:
can these readers be constructed with the schema so you don't need to pass tuples around?

Reply:
It can be done via another refactory. Currently `FileReader` does take the schema. Can make `FileReader` hold the projection schema?

Reply:
it's fine, not a blocker 🤷